keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,776 @@
+  - Estimated increase in size: (e.g., new annotation of size 32B)
+  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
+-->
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+<!--
+Look at the [existing SLIs/SLOs].
+
+Think about adding additional work or introducing new steps in between
+(e.g. need to do X to start a container), etc. Please describe the details.
+
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+-->
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+<!--
+Things to keep in mind include: additional in-memory state, additional
+non-trivial computations, excessive access to disks (including increased log
+volume), significant amount of data sent and/or received over network, etc.
+This through this both in small and large cases, again with respect to the
+[supported limits].
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+-->
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
```

Comment:
One alternative to avoid the validation problem with overlapping ranges, is to have only one resources for the whole cluster that contains an array with the CIDRRanges

```
ClusterCIDR {
  Type      CIDRType
  Ranges []CIDRRange
}

  IPFamily  string # Perhaps omit – inferable from CIDRBlock
  CIDRBlock string # Example "10.0.0.0/8" or "fd12:3456:789a:1::/64"
  Selector  v1.LabelSelector # Specifies which Services or Nodes can be
                               # assigned IPs from this block.
  BlockSize string # How large of an IP block to allocate. For services
                   # this would always be "/32". Example "/24"
```

with cluster CIDRs we don't expect a big number so I don't think the object can grow much

Reply:
Don't we still need to deal with overlapping ranges as the user could specify overlapping CIDRBlocks within the array?

I think overlapping ranges are fine if we coalace all of the allocated ranges in one in memory data structure (maybe a trie) in the backend. In other words, the user can specify overlapping ranges, but the allocator is smart enough to not allocate overlapping ranges.

Reply:
this is related to the commented in the sig-network meeting, right now you don't have access to this information during the API operations.

Reply:
I finally understand what you mean by not having access. We cannot access the backend data structures. Can we access any data structures other than the new object that is being added? Or is the entire validation only on the static content being added?

Reply:
There are two parts here, which makes it a little weird.

Incoming API ops hit admission control and then the REST stack.  We must synchronously validate the data (or not?).  We can define what it means to validate (e.g. just "within a valid range" or "in range and not colliding").  Either way this is async to other operations on CIDRRange resources (e.g. we could validate it and then the range is removed immediately after that - we'd still have an invalid object).

It's not unheard of for admission control to be aware of other objects, but it's not common.  The REST stack is a little weirder.  That's where validation usually happens, and the validation code itself is supposed to be "static".  There are a few places (e.g. Service) that do weird stuff like etcd ops and validating the input is in-range to a flag, so again not unheard of, but uncommon.

The second part - the controller - is all async and much, much easier to reason about.

So IN THEORY we could look at the CIDR ranges during either admission (more self-contained) or validation.  We still have to decide if it is really worthwhile to validate that far, since we can still lose the race.  We don't have interlocks between objects and unless we actually implement referential integrity we can't prevent an object from being deleted or an address from being used twice (apiservers run in parallel)

So the racy path would be:
t0: controller writes node A `podCIDR = 10.10.10.0/24` to apiserver 1
t1: apiserver 1 validates that CIDRange 10.10.0.0/16 allows this allocation AND that no other node is using it
t2: buggy controller writes node B `podCIDR = 10.10.10.0/24` to apiserver 2
t3: apiserver 2 validates that CIDRange 10.10.0.0/16 allows this allocation AND that no other node is using it
t4: apiserver 1 approves the write and it goes to etcd
t5: apiserver 2 approves the write and it goes to etcd

== Nodes A and B have the same CIDR

Another bad path:
t0: controller writes node A `podCIDR = 10.10.10.0/24` to apiserver 1
t1: apiserver 1 validates that CIDRange 10.10.0.0/16 allows this allocation AND that no other node is using it
t2: user deletes the CIDRRange 
t3: apiserver 1 approves the write and it goes to etcd

== Node A has a CIDR that is not covered by any range

To get around the first case we either trust the controller(s) or we force some verifiable contention, causing whomever loses the race to fail the transaction and force the controller to retry.  E.g. the Service bitmap serves this role, but can end up never resolving under extreme contention.  Or you just trust the allocator to have done the right thing (what we do today).

To get around the second case we could use a finalizer - the IPAM controller adds a finalizer like "in-use" to the range it allocates from.  When the user deletes the range, it will hang as long as nodes exist that use the range.  The controller knows that range is being deleted, so it won't allocate more from it.  As nodes are removed, the controller keeps track and when that hits 0, removes the finalizer.  Need to think about what corner cases exist here.

Alternately, implement admission control to reject delete ops while in-use, but I don't think we do that anywhere, so unprecedented.

This still leaves the controller(s) to do safe allocations, which I don't think etcd can help with.  That means we can only have 1 active controller which is normal.  But there is talk of eventually sharding controllers, and this would fail that unless it also shards ranges.

I think this topic warrants more design and pros/cons of approaches.  We may just be making it harder than it needs to be?

Reply:
even more simple, what if we delegate the responsibility to the user, using a Status field set by the controller to say: CIDRRange OK or INVALID

The IPAM controller will just not use the CIDRRanges with invalid status, and it has to be the user to delete it and create a correct one ... is kind of the PV/PVC model, but in  one step only ...

Reply:
I've highlighted in the doc that this controller requires a single writer (and therefore probably some sort of leader-election).

I'll add a section in alternatives considered for the allocation objects being stored, so we have this in our back pocket should the need arise.